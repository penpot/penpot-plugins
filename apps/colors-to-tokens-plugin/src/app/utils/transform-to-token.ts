import { LibraryColor } from '@penpot/plugin-types';
import { TokenStructure } from '../../model';

function transformToRgba({
  color,
  opacity,
}: Required<Pick<LibraryColor, 'color' | 'opacity'>>) {
  color = color.slice(1);

  const r = parseInt(color.substring(0, 2), 16);
  const g = parseInt(color.substring(2, 4), 16);
  const b = parseInt(color.substring(4, 6), 16);

  return `rgba(${r}, ${g}, ${b}, ${opacity})`;
}

interface Color extends LibraryColor {
  color: string;
}

export function transformToToken(colors: LibraryColor[]) {
  const result: TokenStructure = {};

  colors
    .filter((data): data is Color => !!data.color)
    .forEach((data) => {
      const currentOpacity = data.opacity ?? 1;
      const value =
        currentOpacity === 1
          ? data.color
          : transformToRgba({
              opacity: currentOpacity,
              color: data.color,
            });
      const names: string[] = data.name.split(' ');
      const key: string = data.path.replace(' \\/ ', '/').replace(/ /g, '');

      if (!result[key]) {
        result[key] = {};
      }

      const props = [key, ...names];
      let acc = result;

      props.forEach((prop, index) => {
        if (!acc[prop]) {
          acc[prop] = {};
        }

        if (index === props.length - 1) {
          let propIndex = 1;
          const initialProp = prop;

          while (acc[prop]?.$value) {
            prop = `${initialProp}${propIndex}`;
            propIndex++;
          }

          acc[prop] = {
            $value: value,
            $type: 'color',
          };
        }

        acc = acc[prop] as TokenStructure;
      });
    });

  return result;
}
