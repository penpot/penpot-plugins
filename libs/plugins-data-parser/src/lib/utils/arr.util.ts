import { isObject, toCamelCase, isSingleObjectWithProperty } from '.';
import { Arr, Name } from './models/util.model';

/**
 * Checks if "arr" property can be turned into an object
 */
function toObject(arr: unknown): boolean {
  return (
    Array.isArray(arr) && arr.some((a) => isSingleObjectWithProperty(a, 'name'))
  );
}

/**
 * Checks if "arr" property can be turned into an array of objects
 */
function toArray(arr: unknown): boolean {
  return (
    Array.isArray(arr) &&
    arr.every((a) => isObject(a)) &&
    arr.every(
      (a) =>
        isSingleObjectWithProperty(a, 'uuid') ||
        isSingleObjectWithProperty(a, 'arr')
    )
  );
}

/**
 * Parses a splitted "arr" back into an object if possible.
 *
 * If there are objects with a single property "name", that
 * object and the next one are turned into a key-value pair.
 *
 * example:
 * [{ name: 'foo' }, 'bar', { name: 'baz' }, 'qux'] => { foo: 'bar', baz: 'qux' }
 */
function arrToObject(arr: unknown): Record<string, unknown> | null {
  return (
    (Array.isArray(arr) &&
      arr.reduce(
        (result: Record<string, unknown>, value: unknown, index: number) => {
          if (isSingleObjectWithProperty(value, 'name')) {
            const next = arr[index + 1] as unknown;
            if (!!next && !isSingleObjectWithProperty(next, 'name')) {
              return { ...result, [toCamelCase((value as Name)?.name)]: next };
            } else {
              return { ...result, [toCamelCase((value as Name)?.name)]: null };
            }
          }
          return { ...result };
        },
        {}
      )) ||
    null
  );
}

/**
 * Parses a splitted "arr" back into an array of objects.
 *
 */
function arrToArray(arr: unknown): unknown[] | null {
  return (
    (Array.isArray(arr) &&
      arr.reduce((result: unknown[], value: unknown) => {
        if (isSingleObjectWithProperty(value, 'arr')) {
          return [...result, { ...(value as Arr)?.arr }];
        }
        return [...result];
      }, [])) ||
    null
  );
}

/**
 * Checks an "arr" property and decides which parse solution to use
 */
export function parseArrProperty(
  arr: unknown[]
): unknown[] | Record<string, unknown> | null {
  if (toArray(arr)) {
    return arrToArray(arr);
  } else if (toObject(arr)) {
    return arrToObject(arr);
  }

  return arr;
}
