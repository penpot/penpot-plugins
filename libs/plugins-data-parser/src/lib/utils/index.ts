export * from './object.util';
export * from './parse.util';
export * from './selected.util';
export * from './arr.util';
export * from './string.util';
export * from './parse-state';
