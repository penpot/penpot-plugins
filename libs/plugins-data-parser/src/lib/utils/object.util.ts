/**
 * Check if param is an object
 */
export function isObject(obj: unknown): boolean {
  return typeof obj === 'object' && obj !== null && !Array.isArray(obj);
}

/**
 * Check if param is an object with a single property
 */
export function isSingleObjectWithProperty(
  object: unknown,
  property: string
): boolean {
  if (isObject(object)) {
    return (
      Object.keys(object as Record<string, unknown>).length === 1 &&
      !!(object as Record<string, unknown>)[property]
    );
  }

  return false;
}

/**
 * Check if param is an object with a single property from a list
 */
export function isSingleObjectWithProperties(
  object: unknown,
  properties: string[]
): boolean {
  if (isObject(object)) {
    const keys = Object.keys(object as Record<string, unknown>);

    if (keys.length === 1) {
      return properties.includes(keys[0]);
    }
  }

  return false;
}

/**
 * Check if param is a root-tail object
 */
export function isRootTail(obj: unknown): boolean {
  if (isObject(obj)) {
    const keys = Object.keys(obj as Record<string, unknown>);
    return keys.length === 2 && keys.includes('root') && keys.includes('tail');
  }

  return false;
}
