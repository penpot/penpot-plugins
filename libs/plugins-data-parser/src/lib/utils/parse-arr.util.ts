import { Arr, Name, Uuid } from '../models';
import { isObject, toCamelCase } from './object.util';
import { isSingleObjectWithProperty } from './parse-properties.util';

/**
 * Checks if "arr" property can be turned into an object
 */
function toObject(arr: unknown[]): boolean {
  return arr.some((a) => isSingleObjectWithProperty(a, 'name'));
}

/**
 * Checks if "arr" property can be turned into an array of objects
 */
function toArray(arr: unknown[]): boolean {
  return (
    arr.every((a) => isObject(a)) &&
    arr.some((a) => isSingleObjectWithProperty(a, 'uuid')) &&
    arr.some((a) => isSingleObjectWithProperty(a, 'arr'))
  );
}

/**
 * Checks if "arr" needs cleaning and clean the leftovers uuid objects.
 *
 * It needs cleaning when uuid objects are redundant.
 */
function cleanUuid(arr: unknown[]): unknown[] {
  const shouldClean = arr.some((a, index) => {
    const next = arr[index + 1] as Record<string, unknown>;

    return (
      isSingleObjectWithProperty(a, 'uuid') &&
      (next?.['id'] as Uuid)?.uuid === (a as Uuid).uuid
    );
  });

  if (shouldClean) {
    return arr.filter((a) => !isSingleObjectWithProperty(a, 'uuid'));
  }

  return arr;
}

/**
 * Parses a splitted "arr" back into an object if possible.
 *
 * If there are objects with a single property "name", that
 * object and the next one are turned into a key-value pair.
 *
 * example:
 * [{ name: 'foo' }, 'bar', { name: 'baz' }, 'qux'] => { foo: 'bar', baz: 'qux' }
 */
function arrToObject(arr: unknown[]): Record<string, unknown> {
  return arr.reduce(
    (result: Record<string, unknown>, value: unknown, index: number) => {
      if (isSingleObjectWithProperty(value, 'name')) {
        const next = arr[index + 1];
        if (!!next && !isSingleObjectWithProperty(next, 'name')) {
          return { ...result, [toCamelCase((value as Name).name)]: next };
        } else {
          return { ...result, [toCamelCase((value as Name).name)]: null };
        }
      }
      return { ...result };
    },
    {}
  );
}

/**
 * Recursively parses a splitted "arr" back into an array of
 * objects with id and data properties.
 *
 * If there are objects with a single property "uuid", and
 * the next one is an object with a single property "arr",
 * it turns them into an object with id and data properties.
 *
 * It also checks for nested "arr" properties and turns them
 * into an object with key-value pairs if possible.
 *
 * example:
 *
 * [{ uuid: 'foo' }, {arr: [{ name: 'bar' }, 'baz']}] => [{ id: 'foo', data: { bar: 'baz' } }]
 */
function arrToArray(arr: unknown[]): unknown[] {
  return arr.reduce((result: unknown[], value: unknown, index: number) => {
    if (isSingleObjectWithProperty(value, 'uuid')) {
      const next = arr[index + 1];
      if (!!next && isSingleObjectWithProperty(next, 'arr')) {
        const parsedArr = toObject((next as Arr).arr)
          ? arrToObject((next as Arr).arr)
          : toArray((next as Arr).arr)
          ? arrToArray((next as Arr).arr)
          : [...(next as Arr).arr];

        return [...result, { ...parsedArr }];
      }
    }
    return [...result];
  }, []);
}

/**
 * Checks an "arr" property and decides which parse solution to use
 */
export function parseArrProperty(
  arr: unknown[]
): unknown[] | Record<string, unknown> {
  if (toArray(arr)) {
    return arrToArray(arr);
  } else if (toObject(arr)) {
    return arrToObject(arr);
  }

  return cleanUuid(arr);
}

/**
 * Parses an object with an "arr" property
 */
export function parseObjArr(obj: unknown): unknown {
  if (isSingleObjectWithProperty(obj, 'arr')) {
    return parseArrProperty((obj as Arr)['arr']);
  }

  return obj;
}

/**
 * Checks if an array is a nested array of objects
 *
 * It also checks and filter empty nested arrays
 */
function isNestedArray(arr: unknown[]): boolean {
  if (Array.isArray(arr) && arr.every((a) => Array.isArray(a))) {
    // Filter empty nested arrays
    const filtered = arr.filter((a) => (a as unknown[]).length > 0);

    // Check if every nested array is an array of objects
    return filtered.every(
      (a) => Array.isArray(a) && a.every((b) => isObject(b))
    );
  }

  return false;
}

/**
 * If an array is nested, it flattens it.
 *
 * example
 * [[1, 2], [3, 4]] => [1, 2, 3, 4]
 */
export function flattenNestedArrays(arr: unknown[]): unknown {
  if (isNestedArray(arr)) {
    return arr.flatMap((innerArray) => innerArray);
  }
  return arr;
}
