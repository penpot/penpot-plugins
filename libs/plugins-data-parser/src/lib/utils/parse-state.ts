/* eslint-disable */

import { toCamelCase } from './object.util';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type UknowCljs = any;

declare global {
  const cljs: UknowCljs;
  const app: UknowCljs;
}

const PATH_SEPARATOR = '/' as const;

function getPath(path: string): UknowCljs[] {
  if (path === 'root') {
    return [];
  }
  return path
    .slice(5)
    .split(PATH_SEPARATOR)
    .map((it) => {
      if (it.startsWith('#')) {
        return cljs.core.uuid(it.slice(1));
      } else if (it.startsWith(':')) {
        return cljs.core.keyword(it.slice(1));
      } else if (it.startsWith('!')) {
        return parseInt(it.slice(1), 10);
      } else {
        return it;
      }
    });
}

const expandableTable = {
  'app:datetime': false,
  'cljs:uuid': false,
  'cljs:keyword': false,
  'cljs:symbol': false,
  'js:date': false,
  'js:boolean': false,
  'js:function': false,
  'js:number': false,
  'js:string': false,
  'js:undefined': false,
  'js:null': false,
  'unkwnown:unknown': false,
  'cljs:vector': true,
  'cljs:list': true,
  'cljs:map': true,
  'cljs:set': true,
  'js:array': true,
  'js:object': true,
} as const;

function isCljsArray(key: string) {
  const valid = ['cljs:vector', 'cljs:list', 'cljs:set'];

  return valid.includes(key);
}

const valueTable = {
  'app:datetime': (value: unknown) => cljs.core.str(value),
  'cljs:uuid': (value: unknown) => cljs.core.name(value),
  'cljs:keyword': (value: unknown) => cljs.core.name(value),
  'js:date': (value: unknown) => (value as Date).toISOString(),
};

function isTypeExpandable(type: keyof typeof expandableTable) {
  return expandableTable[type];
}

function getType(value: unknown) {
  const typeTable = {
    'app:datetime': app.util.time.datetime_QMARK_,
    'cljs:uuid': cljs.core.uuid_QMARK_,
    'cljs:keyword': cljs.core.keyword_QMARK_,
    'cljs:vector': cljs.core.vector_QMARK_,
    'cljs:list': cljs.core.list_QMARK_,
    'cljs:map': cljs.core.map_QMARK_,
    'cljs:set': cljs.core.set_QMARK_,
    'cljs:symbol': cljs.core.symbol_QMARK_,
    'js:date': (value: unknown) => value instanceof Date,
    'js:array': (value: unknown) => Array.isArray(value),
    'js:boolean': (value: unknown) => typeof value === 'boolean',
    'js:function': (value: unknown) => typeof value === 'function',
    'js:number': (value: unknown) => typeof value === 'number',
    'js:string': (value: unknown) => typeof value === 'string',
    'js:undefined': (value: unknown) => typeof value === 'undefined',
    'js:null': (value: unknown) => value === null,
    'js:object': (value: unknown) =>
      typeof value === 'object' &&
      value !== null &&
      value.toString() === '[object Object]',
  };

  for (const [type, fn] of Object.entries(typeTable)) {
    if (fn(value)) return type as keyof typeof typeTable;
  }

  return 'unknown:unknown' as keyof typeof typeTable;
}

function getKeyName(key: string | number) {
  if (cljs.core.uuid_QMARK_(key)) {
    return '#' + cljs.core.name(key);
  } else if (cljs.core.keyword_QMARK_(key)) {
    return ':' + cljs.core.name(key);
  } else if (typeof key === 'string') {
    return key;
  } else if (typeof key === 'number') {
    return `!${key}`;
  }

  return key;
}

function getValue(value: unknown, type: string): unknown {
  return type in valueTable
    ? valueTable[type as keyof typeof valueTable](value)
    : value;
}

export function getPartialState(path: string, state: UknowCljs): UknowCljs {
  // const state = cljs.core.deref(app.main.store.state);
  const statePath = getPath(path);
  const data = cljs.core.get_in(state, statePath);
  const type = getType(data);
  const isArray = isCljsArray(type);
  const isExpandable = isTypeExpandable(type);

  const map: Record<string, unknown> = {};
  const list: unknown[] = [];

  function addEntry(
    childPath: string,
    propName: string | number,
    repValue: unknown,
    isExpandable: boolean
  ) {
    if (isArray) {
      list.push(isExpandable ? getPartialState(childPath, state) : repValue);
    } else {
      if (isExpandable) {
        Object.defineProperty(map, propName, {
          get() {
            return getPartialState(childPath, state);
          },
          configurable: true,
          enumerable: true,
        });
      } else {
        map[propName] = repValue;
      }
    }
  }

  if (isExpandable) {
    if (type.startsWith('cljs:')) {
      const cljsData = cljs.core.clj__GT_js(cljs.core.to_array(data));

      cljsData.forEach((entry: unknown, index: number) => {
        if (cljs.core.map_entry_QMARK_(entry)) {
          const key = cljs.core.first(entry);
          const name = getKeyName(key);
          const value = cljs.core.nth(entry, 1);
          const valueType = getType(value);
          const repValue = getValue(value, valueType);
          const isExpandable = isTypeExpandable(valueType);
          const childPath = path + PATH_SEPARATOR + name;
          const propName = name.startsWith(':')
            ? toCamelCase(name.slice(1))
            : name;

          addEntry(childPath, propName, repValue, isExpandable);
        } else {
          const valueType = getType(entry);
          const repValue = getValue(entry, valueType);
          const isExpandable = isTypeExpandable(valueType);
          const childPath = path + PATH_SEPARATOR + getKeyName(index);

          addEntry(childPath, index, repValue, isExpandable);
        }
      });
    } else if (type.startsWith('js:')) {
      Object.entries(data).forEach(([key, value]) => {
        map[key] = value;
      });
    }

    return isArray ? list : map;
  }

  const repValue = getValue(data, type);

  return repValue;
}
