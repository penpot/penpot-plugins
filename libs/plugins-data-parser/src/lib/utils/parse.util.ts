import {
  isObject,
  toCamelCase,
  parseArrProperty,
  isRootTail,
  isSingleObjectWithProperties,
  isSingleObjectWithProperty,
} from '.';
import { Arr } from './models/util.model';

export function parseSingleProperties(
  obj: unknown,
  properties: string[]
): unknown {
  let result = obj;

  properties.forEach((property) => {
    if (isSingleObjectWithProperty(obj, property)) {
      result = (obj as Record<string, unknown>)[property];
    }
  });

  return result;
}

export function parseRootTail(obj: unknown): unknown {
  if (isObject(obj) && isRootTail(obj)) {
    const { root, tail } = obj as Record<string, unknown>;

    const hasRoot = Array.isArray(root) && root?.length > 0;
    const hasTail = Array.isArray(tail) && tail?.length > 0;

    if (hasRoot && hasTail) {
      return obj;
    }

    if (hasTail) {
      return tail;
    }

    if (hasRoot) {
      return root;
    }

    return [];
  }

  return obj;
}

/**
 * Recursively cleans an object from unnecesary properties
 * and converts snake_case and kebab-case to camelCase
 */
export function cleanObject(obj: unknown): unknown {
  if (Array.isArray(obj)) {
    return obj
      .filter((p) => p !== null)
      .map((v: Record<string, unknown>) => cleanObject(v)) as Record<
      string,
      unknown
    >[];
  } else if (isObject(obj)) {
    return Object.keys(obj as Record<string, unknown>)
      .filter(
        (key) =>
          !/^(cljs\$|\$hash|ts|ry|\$meta|__hash|_hash|bitmap|meta|ns|fqn|cnt|shift|edit|has_nil_QMARK_|nil_val)/g.test(
            key
          )
      )
      .reduce((result, key) => {
        const value = (obj as Record<string, unknown>)[key];
        if (['extmap', '$extmap'].includes(key) && value === null) {
          return { ...result };
        }

        return {
          ...result,
          [toCamelCase(key)]: cleanObject(
            (obj as Record<string, unknown>)[key]
          ),
        };
      }, {});
  }
  return obj as Record<string, unknown>;
}

export function parseObject(obj: unknown): unknown {
  const singleProperties = ['root', 'name', 'uuid', 'guides'];
  if (isSingleObjectWithProperties(obj, singleProperties)) {
    const parsed = parseSingleProperties(
      obj as Record<string, unknown>,
      singleProperties
    );
    return parseObject(parsed);
  }

  if (isSingleObjectWithProperty(obj, 'arr')) {
    const parsed = parseArrProperty((obj as Arr).arr);
    return parseObject(parsed);
  }

  if (isRootTail(obj)) {
    const parsed = parseRootTail(obj);
    return parseObject(parsed);
  }

  // If it's an array, parse each element
  if (Array.isArray(obj)) {
    return obj.map((v: Record<string, unknown>) => parseObject(v));
  }

  // If it's an object, parse each property
  if (isObject(obj)) {
    return Object.keys(obj as Record<string, unknown>).reduce((result, key) => {
      return {
        ...result,
        [key]: parseObject((obj as Record<string, unknown>)[key]),
      };
    }, {});
  }

  return obj;
}
