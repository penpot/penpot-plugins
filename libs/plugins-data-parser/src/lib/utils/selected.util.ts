import { UnparsedSelection } from '../models/selection.model';

/**
 * Gets selected uuids from selection object
 */
export function getSelectedUuids(selection: UnparsedSelection): string[] {
  const root = selection?.linked_map?.delegate_map?.root?.arr;

  if (!root) {
    return [];
  }

  return root.map((r) => r.uuid).filter((uuid): uuid is string => !!uuid);
}
