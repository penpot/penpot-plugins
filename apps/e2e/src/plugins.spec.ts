import testingPlugin from './plugins/create-frame-text-rect';
import { Agent } from './utils/agent';

describe('Plugins', () => {
  it('create frame - text - rectable', async () => {
    const agent = await Agent();
    const result = await agent.runCode(testingPlugin.toString());

    expect(result).toMatchSnapshot();
  });
});
