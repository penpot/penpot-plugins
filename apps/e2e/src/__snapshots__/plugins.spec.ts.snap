// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Plugins > create frame - text - rectable 1`] = `
[
  {
    ":fills": [
      {
        ":fill-color": "#FFFFFF",
        ":fill-opacity": 1,
      },
    ],
    ":flip-x": null,
    ":flip-y": null,
    ":frame-id": "1",
    ":height": 0.01,
    ":hide-fill-on-export": false,
    ":id": "1",
    ":name": "Root Frame",
    ":parent-id": "1",
    ":points": [
      {
        ":x": 0,
        ":y": 0,
      },
      {
        ":x": 0.01,
        ":y": 0,
      },
      {
        ":x": 0.01,
        ":y": 0.01,
      },
      {
        ":x": 0,
        ":y": 0.01,
      },
    ],
    ":proportion": 1,
    ":proportion-lock": false,
    ":rotation": 0,
    ":selrect": {
      ":height": 0.01,
      ":width": 0.01,
      ":x": 0,
      ":x1": 0,
      ":x2": 0.01,
      ":y": 0,
      ":y1": 0,
      ":y2": 0.01,
    },
    ":shapes": [
      "2",
      "3",
      "4",
    ],
    ":strokes": [],
    ":transform": {
      ":a": 1,
      ":b": 0,
      ":c": 0,
      ":d": 1,
      ":e": 0,
      ":f": 0,
    },
    ":transform-inverse": {
      ":a": 1,
      ":b": 0,
      ":c": 0,
      ":d": 1,
      ":e": 0,
      ":f": 0,
    },
    ":type": ":frame",
    ":width": 0.01,
    ":x": 0,
    ":y": 0,
  },
  {
    ":content": {
      ":children": [
        {
          ":children": [
            {
              ":children": [
                {
                  ":fills": [
                    {
                      ":fill-color": "#000000",
                      ":fill-opacity": 1,
                    },
                  ],
                  ":font-family": "sourcesanspro",
                  ":font-id": "sourcesanspro",
                  ":font-size": "14",
                  ":font-style": "normal",
                  ":font-variant-id": "regular",
                  ":font-weight": "400",
                  ":letter-spacing": "0",
                  ":line-height": "1.2",
                  ":text": "Hello from plugin",
                  ":text-align": "left",
                  ":text-decoration": "none",
                  ":text-transform": "none",
                  ":typography-ref-file": null,
                  ":typography-ref-id": null,
                },
              ],
              ":fills": [
                {
                  ":fill-color": "#000000",
                  ":fill-opacity": 1,
                },
              ],
              ":font-family": "sourcesanspro",
              ":font-id": "sourcesanspro",
              ":font-size": "14",
              ":font-style": "normal",
              ":font-variant-id": "regular",
              ":font-weight": "400",
              ":letter-spacing": "0",
              ":line-height": "1.2",
              ":text-align": "left",
              ":text-decoration": "none",
              ":text-transform": "none",
              ":type": "paragraph",
              ":typography-ref-file": null,
              ":typography-ref-id": null,
            },
          ],
          ":type": "paragraph-set",
        },
      ],
      ":type": "root",
    },
    ":flip-x": null,
    ":flip-y": null,
    ":frame-id": "1",
    ":grow-type": ":auto-width",
    ":height": 17,
    ":id": "2",
    ":name": "Text",
    ":parent-id": "1",
    ":points": [
      {
        ":x": 684,
        ":y": 540,
      },
      {
        ":x": 786,
        ":y": 540,
      },
      {
        ":x": 786,
        ":y": 557,
      },
      {
        ":x": 684,
        ":y": 557,
      },
    ],
    ":position-data": [
      {
        ":direction": "ltr",
        ":fills": [
          {
            ":fill-color": "#000000",
            ":fill-opacity": 1,
          },
        ],
        ":font-family": "sourcesanspro",
        ":font-size": "14px",
        ":font-style": "normal",
        ":font-weight": "400",
        ":height": 18,
        ":letter-spacing": "normal",
        ":text": "Hello from plugin",
        ":text-decoration": "none solid rgb(0, 0, 0)",
        ":text-transform": "none",
        ":width": 101.5313,
        ":x": 684,
        ":x1": 0,
        ":x2": 101.5313,
        ":y": 557,
        ":y1": -1,
        ":y2": 17,
      },
    ],
    ":rotation": 0,
    ":selrect": {
      ":height": 17,
      ":width": 102,
      ":x": 684,
      ":x1": 684,
      ":x2": 786,
      ":y": 540,
      ":y1": 540,
      ":y2": 557,
    },
    ":transform": {
      ":a": 1,
      ":b": 0,
      ":c": 0,
      ":d": 1,
      ":e": 0,
      ":f": 0,
    },
    ":transform-inverse": {
      ":a": 1,
      ":b": 0,
      ":c": 0,
      ":d": 1,
      ":e": 0,
      ":f": 0,
    },
    ":type": ":text",
    ":width": 102,
    ":x": 684,
    ":y": 540,
  },
  {
    ":fills": [
      {
        ":fill-color": "#B1B2B5",
        ":fill-opacity": 1,
      },
    ],
    ":flip-x": null,
    ":flip-y": null,
    ":frame-id": "1",
    ":height": 200,
    ":id": "3",
    ":name": "Rectangle",
    ":parent-id": "1",
    ":plugin-data": {
      ":plugin/TEST": {
        "customKey": "customValue",
      },
    },
    ":points": [
      {
        ":x": 684,
        ":y": 540,
      },
      {
        ":x": 884,
        ":y": 540,
      },
      {
        ":x": 884,
        ":y": 740,
      },
      {
        ":x": 684,
        ":y": 740,
      },
    ],
    ":proportion": 1,
    ":proportion-lock": false,
    ":rotation": 0,
    ":rx": 0,
    ":ry": 0,
    ":selrect": {
      ":height": 200,
      ":width": 200,
      ":x": 684,
      ":x1": 684,
      ":x2": 884,
      ":y": 540,
      ":y1": 540,
      ":y2": 740,
    },
    ":strokes": [],
    ":transform": {
      ":a": 1,
      ":b": 0,
      ":c": 0,
      ":d": 1,
      ":e": 0,
      ":f": 0,
    },
    ":transform-inverse": {
      ":a": 1,
      ":b": 0,
      ":c": 0,
      ":d": 1,
      ":e": 0,
      ":f": 0,
    },
    ":type": ":rect",
    ":width": 200,
    ":x": 684,
    ":y": 540,
  },
  {
    ":fills": [
      {
        ":fill-color": "#FFFFFF",
        ":fill-opacity": 1,
      },
    ],
    ":flip-x": null,
    ":flip-y": null,
    ":frame-id": "1",
    ":height": 300,
    ":hide-fill-on-export": false,
    ":id": "4",
    ":name": "Frame name",
    ":parent-id": "1",
    ":points": [
      {
        ":x": 684,
        ":y": 540,
      },
      {
        ":x": 984,
        ":y": 540,
      },
      {
        ":x": 984,
        ":y": 840,
      },
      {
        ":x": 684,
        ":y": 840,
      },
    ],
    ":proportion": 1,
    ":proportion-lock": false,
    ":rotation": 0,
    ":rx": 8,
    ":ry": 8,
    ":selrect": {
      ":height": 300,
      ":width": 300,
      ":x": 684,
      ":x1": 684,
      ":x2": 984,
      ":y": 540,
      ":y1": 540,
      ":y2": 840,
    },
    ":shapes": [
      "5",
    ],
    ":strokes": [],
    ":transform": {
      ":a": 1,
      ":b": 0,
      ":c": 0,
      ":d": 1,
      ":e": 0,
      ":f": 0,
    },
    ":transform-inverse": {
      ":a": 1,
      ":b": 0,
      ":c": 0,
      ":d": 1,
      ":e": 0,
      ":f": 0,
    },
    ":type": ":frame",
    ":width": 300,
    ":x": 684,
    ":y": 540,
  },
  {
    ":constraints-h": ":left",
    ":constraints-v": ":top",
    ":content": {
      ":children": [
        {
          ":children": [
            {
              ":children": [
                {
                  ":fills": [
                    {
                      ":fill-color": "#000000",
                      ":fill-opacity": 1,
                    },
                  ],
                  ":font-family": "sourcesanspro",
                  ":font-id": "sourcesanspro",
                  ":font-size": "14",
                  ":font-style": "normal",
                  ":font-variant-id": "regular",
                  ":font-weight": "400",
                  ":letter-spacing": "0",
                  ":line-height": "1.2",
                  ":text": "Hello from frame",
                  ":text-align": "left",
                  ":text-decoration": "none",
                  ":text-transform": "none",
                  ":typography-ref-file": null,
                  ":typography-ref-id": null,
                },
              ],
              ":fills": [
                {
                  ":fill-color": "#000000",
                  ":fill-opacity": 1,
                },
              ],
              ":font-family": "sourcesanspro",
              ":font-id": "sourcesanspro",
              ":font-size": "14",
              ":font-style": "normal",
              ":font-variant-id": "regular",
              ":font-weight": "400",
              ":letter-spacing": "0",
              ":line-height": "1.2",
              ":text-align": "left",
              ":text-decoration": "none",
              ":text-transform": "none",
              ":type": "paragraph",
              ":typography-ref-file": null,
              ":typography-ref-id": null,
            },
          ],
          ":type": "paragraph-set",
        },
      ],
      ":type": "root",
    },
    ":flip-x": null,
    ":flip-y": null,
    ":frame-id": "4",
    ":grow-type": ":auto-width",
    ":height": 17,
    ":id": "5",
    ":name": "Text",
    ":parent-id": "4",
    ":points": [
      {
        ":x": 10,
        ":y": 10,
      },
      {
        ":x": 109,
        ":y": 10,
      },
      {
        ":x": 109,
        ":y": 27,
      },
      {
        ":x": 10,
        ":y": 27,
      },
    ],
    ":position-data": [
      {
        ":direction": "ltr",
        ":fills": [
          {
            ":fill-color": "#000000",
            ":fill-opacity": 1,
          },
        ],
        ":font-family": "sourcesanspro",
        ":font-size": "14px",
        ":font-style": "normal",
        ":font-weight": "400",
        ":height": 18,
        ":letter-spacing": "normal",
        ":text": "Hello from frame",
        ":text-decoration": "none solid rgb(0, 0, 0)",
        ":text-transform": "none",
        ":width": 98.7344,
        ":x": 10,
        ":x1": 0,
        ":x2": 98.7344,
        ":y": 27,
        ":y1": -1,
        ":y2": 17,
      },
    ],
    ":rotation": 0,
    ":selrect": {
      ":height": 17,
      ":width": 99,
      ":x": 10,
      ":x1": 10,
      ":x2": 109,
      ":y": 10,
      ":y1": 10,
      ":y2": 27,
    },
    ":transform": {
      ":a": 1,
      ":b": 0,
      ":c": 0,
      ":d": 1,
      ":e": 0,
      ":f": 0,
    },
    ":transform-inverse": {
      ":a": 1,
      ":b": 0,
      ":c": 0,
      ":d": 1,
      ":e": 0,
      ":f": 0,
    },
    ":type": ":text",
    ":width": 99,
    ":x": 10,
    ":y": 10,
  },
]
`;
